src/Compilers/CSharp/Test/Emit3/FirstClassSpanTests.cs

```diff
@@ -7408,4 +7543,282 @@ static class C
+    }
+
+    [Fact]
+    public void TypeInference_Span_Array()
+    {
+        var source = """
+            using System;
+            class C
+            {
+                void M(Span<int> a) => M1(a);
+                void M1<T>(T[] x) { }
+            }
+            """;
+        CreateCompilationWithSpanAndMemoryExtensions(source).VerifyDiagnostics(
+            // (4,28): error CS0411: The type arguments for method 'C.M1<T>(T[])' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+            //     void M(Span<int> a) => M1(a);
+            Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "M1").WithArguments("C.M1<T>(T[])").WithLocation(4, 28));
+    }
+
+    [Fact]
+    public void TypeInference_Array_ReadOnlySpan_01()
+    {
+        var source = """
+            using System;
+            class C
+            {
+                void M(int[] a) => M1(a);
+                void M1<T>(ReadOnlySpan<T> x) { }
+            }
+            """;
+        var comp = CreateCompilationWithSpanAndMemoryExtensions(source).VerifyDiagnostics();
+        AssertEx.Equal("C.M1<System.Int32>", DisplayInvokedMethodTypeArguments(comp));
+    }
+
+    [Fact]
+    public void TypeInference_Array_ReadOnlySpan_02()
+    {
+        var source = """
+            using System;
+            class C
+            {
+                void M(string[] a, object b) => M1(a, b);
+                void M1<T>(ReadOnlySpan<T> x, T y) { }
+            }
+            """;
+        var comp = CreateCompilationWithSpanAndMemoryExtensions(source).VerifyDiagnostics();
+        AssertEx.Equal("C.M1<System.Object>", DisplayInvokedMethodTypeArguments(comp));
+    }
+
+    [Fact]
+    public void TypeInference_Array_ReadOnlySpan_03()
+    {
+        var source = """
+            using System;
+            class C
+            {
+                void M(object[] a, string b) => M1(a, b);
+                void M1<T>(ReadOnlySpan<T> x, T y) { }
+            }
+            """;
+        var comp = CreateCompilationWithSpanAndMemoryExtensions(source).VerifyDiagnostics();
+        AssertEx.Equal("C.M1<System.Object>", DisplayInvokedMethodTypeArguments(comp));
+    }
+
+    [Fact]
+    public void TypeInference_ReadOnlySpan_ReadOnlySpan_01()
+    {
+        var source = """
+            using System;
+            class C
+            {
+                void M(ReadOnlySpan<int> a) => M1(a);
+                void M1<T>(ReadOnlySpan<T> x) { }
+            }
+            """;
+        var comp = CreateCompilationWithSpanAndMemoryExtensions(source).VerifyDiagnostics();
+        AssertEx.Equal("C.M1<System.Int32>", DisplayInvokedMethodTypeArguments(comp));
+    }
+
+    [Fact]
+    public void TypeInference_ReadOnlySpan_ReadOnlySpan_02()
+    {
+        var source = """
+            using System;
+            class C
+            {
+                void M(ReadOnlySpan<string> a, object b) => M1(a, b);
```

Comment:
Let's test a scenario like this with a non-variant non-array conversion, like `ROS<int>` and `object`/vice versa; I expect the former to fail, and the latter to succeed.